sycl/doc/design/CommandGraph.md

```diff
@@ -1,66 +1,164 @@
 
 There is no extra support for Graph-specific USM allocations in the current
 proposal. Memory operations will be supported subsequently by the current
 implementation starting with `memcpy`.
 
-Buffers and accessors are supported in a command graph. Following restrictions
+Buffers and accessors are supported in a command-graph. Following restrictions
 are required to adapt buffers and their lifetime to a lazy work execution model:
 
 - Lifetime of a buffer with host data will be extended by copying the underlying
 data.
-- Host accessor on buffer that are used by a command graph are prohibited.
-- Copy-back behavior on destruction of a buffer is prohibited. 
+- Host accessor on buffer that are used by a command-graph are prohibited.
+- Copy-back behavior on destruction of a buffer is prohibited.
+
+## Backend Implementation
+
+Implementation of [UR command-buffers](#UR-command-buffer-experimental-feature)
+for each of the supported SYCL 2020 backends.
+
+### Level Zero
+
+The UR `urCommandBufferEnqueueExp` interface for submitting a command-buffer
+takes a list of events to wait on, and returns an event representing the
+completion of that specific submission of the command-buffer.
+
+However, in the equivalent Level Zero function
+[zeCommandQueueExecuteCommandLists](https://spec.oneapi.io/level-zero/latest/core/api.html#zecommandqueueexecutecommandlists)
+there are no parameters to take a wait-list, and the only sync primitive
+returned is blocking on host.
+
+In order to get the UR command-buffer enqueue semantics we want with Level Zero,
+the adapter implementation adds extra commands to the Level Zero command-list
+representing a UR command-buffer.
+
+* Prefix - Commands added to the start of the L0 command-list by L0 adapter.
+* Suffix - Commands added to the end of the L0 command-list by L0 adapter.
+
+These extra commands operate on L0 event synchronisation primitives, used by the
+command-list to interact with the external UR wait-list and UR return event
+required for the enqueue interface.
+
+The `ur_exp_command_buffer_handle_t` class for this adapter contains a
+*SignalEvent* which signals the completion of the command-list in the suffix,
+and is reset in the prefix. This signal is detected by a new UR return event
+created on UR command-buffer enqueue.
+
+There is also a *WaitEvent* used by the `ur_exp_command_buffer_handle_t` class
+in the prefix to wait on any dependencies passed in the enqueue wait-list.
+
+```mermaid
+flowchart TB
+    subgraph Suffix
+    id1[Signal the UR command-buffer SignalEvent]
+    end
+    subgraph User Added Commands
+    id2[Command 1] -.-  id3[Command N]
+    end
+    subgraph Prefix
+    id4[Reset SignalEvent] ~~~ id5[Barrier waiting on WaitEvent]
+    end
+```
+
+For a call to `urCommandBufferEnqueueExp` with an `event_list` *EL*,
+command-buffer *CB*, and return event *RE* our implementation has to create and
+submit two new command-lists for the above approach to work. One before
```

Comment:
I don't know if this is too nitpicky but this might be misleading a reader to think that we actually create a new command list manually every time for these when we are getting them from the UR Queue which will likely just reuse one, and only create a new one in the worst case.

Reply:
Rephrased this to mention getting new command-lists from UR queue